services/api/service.go

```diff
@@ -493,6 +527,84 @@ func (api *RelayAPI) startValidatorRegistrationDBProcessor() {
 	}
 }
 
+// simulateBlock sends a request for a block simulation to blockSimRateLimiter.
+func (api *RelayAPI) simulateBlock(ctx context.Context, opts blockSimOptions) (error, error) {
+	t := time.Now()
+	reqErr, simErr := api.blockSimRateLimiter.Send(ctx, opts.req, opts.isHighPrio, opts.fastTrack)
+	log := opts.log.WithFields(logrus.Fields{
+		"duration":   time.Since(t).Seconds(),
+		"numWaiting": api.blockSimRateLimiter.CurrentCounter(),
+	})
+	if simErr != nil &&
+		simErr.Error() != ErrBlockAlreadyKnown &&
+		simErr.Error() != ErrBlockRequiresReorg &&
+		!strings.Contains(simErr.Error(), ErrMissingTrieNode) {
+		// Mark builder as non-optimistic.
+		opts.builder.status.IsOptimistic = false
+		log.WithError(simErr).Error("block validation failed")
+		return nil, simErr
+	}
+	if reqErr != nil {
+		log.WithError(reqErr).Error("block validation failed: request error")
+		return reqErr, nil
+	}
+	log.Info("block validation successful")
+	return nil, nil
+}
+
+func (api *RelayAPI) demoteBuilder(pubkey string, req *common.BuilderSubmitBlockRequest, simError error) {
+	builderEntry, ok := api.blockBuildersCache[pubkey]
+	if !ok {
+		api.log.Warnf("builder %v not in the builder cache", pubkey)
+		builderEntry = &blockBuilderCacheEntry{} //nolint:exhaustruct
+	}
+	newStatus := common.BuilderStatus{
+		IsHighPrio:    builderEntry.status.IsHighPrio,
+		IsBlacklisted: builderEntry.status.IsBlacklisted,
+		IsOptimistic:  false,
+	}
+	api.log.Infof("demoted builder, new status: %v", newStatus)
+	if err := api.db.SetBlockBuilderStatus(pubkey, newStatus, true); err != nil {
+		api.log.Error(fmt.Errorf("error setting builder: %v status: %w", pubkey, err))
+	}
+	// Write to demotions table.
+	api.log.WithFields(logrus.Fields{"builder_pubkey": pubkey}).Info("demoting builder")
+	if err := api.db.InsertBuilderDemotion(req, simError); err != nil {
+		api.log.WithError(err).WithFields(logrus.Fields{
+			"errorWritingDemotionToDB": true,
+			"bidTrace":                 req.Message,
+			"simError":                 simError,
+		}).Error("failed to save demotion to database")
+	}
+}
+
+// processOptimisticBlock is called on a new goroutine when a optimistic block
+// needs to be simulated.
+func (api *RelayAPI) processOptimisticBlock(opts blockSimOptions) {
```

Comment:
not immediately necessary, but should we add a way to the prio-load-balancer to put such a validation request on top of the fast-prio queue? (so it get's validated as quickly as possible, or does it actually not matter)

Reply:
we actually dont care about optimistic blocks being validated quickly. we dont want to interfere with the other blocks that have to be validated before they are made eligible. so we intentionally put the optimistic blocks on the low-prio queue.